### ISSUE TITLE
`extra_unused_type_parameters` fires with `#![allow(unused)]`

### ISSUE THREAD

**Issue by tgross35** [rust-lang/rust-clippy#12385]
### Summary

`extra_unused_type_parameters` still fires with `#![allow(unused)]` or `dead_code` set. It should probably not fire in these cases.

### Lint Name

extra_unused_type_parameters

### Reproducer

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=3e9fab384e5c8c15f10b55ee39965114

I tried this code:

```rust
#![allow(unused)]
#![warn(clippy::extra_unused_type_parameters)]

fn foo<T: Clone>() {
    todo!()
}
```

I saw this happen:

```
warning: type parameter `T` goes unused in function definition
 --> src/lib.rs:4:7
  |
4 | fn foo<T: Clone>() {
  |       ^^^^^^^^^^ help: consider removing the parameter
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#extra_unused_type_parameters
note: the lint level is defined here
 --> src/lib.rs:2:9
  |
2 | #![warn(clippy::extra_unused_type_parameters)]
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

I expected to see this happen: No error

### Version

```text
Current playground, 1.76.0
```
